docs/dev/design/new-search-api.rst

```diff
@@ -0,0 +1,298 @@
+  instead of having the options in the database.
+- Allow to search a group of projects/versions at the same time.
+- Bring the same syntax to the dashboard search.
+
+Syntax
+------
+
+The parameters will be given in the query using the ``key:value`` syntax.
+Inspired by `GitHub <https://docs.github.com/en/rest/search>`__ and other services.
+
+Currently the values from all parameters don't include spaces,
+so surrounding the value with quotes won't be supported (``key:"value"``).
+
+To avoid interpreting a query as a parameter,
+an escape character can be put in place,
+for example ``project\:docs`` won't be interpreted as
+a parameter, but as the search term ``project:docs``.
+This is only necessary if the query includes a valid parameter,
+unknown parameters (``foo:bar``) don't require escaping.
+
+All other tokens that don't match a valid parameter,
+will be join to form the final search term.
+
+Parameters
+----------
+
+project:
+ Indicates the project and version
+ to includes results from (this doesn't include subprojects).
+ If the version isn't provided,
+ the default version is used.
+
+ Examples:
+
+ - ``project:docs/latest``
+ - ``project:docs``
+
+ It can be one or more project parameters.
+ At least one is required.
+
+ If the user doesn't have permission over one version or if the version doesn't exist,
+ we don't include results from that version.
+ We don't fail the search, this is so users can use one endpoint for all their users,
+ without worrying about what permissions each user has or updating it after a version or project
+ has been deleted.
+
+ The ``/`` is used as separator,
+ but it could be any other character that isn't present in the slug of a version or project.
+ ``:`` was considered (``project:docs:latest``), but it could be hard to read
+ since ``:`` is already used to separate the key from the value.
+
+Including subprojects
+~~~~~~~~~~~~~~~~~~~~~
+
+Now that we are returning results only
+from the given projects, we need an easy way to
+include results from subprojects.
+Some ideas for implementing this feature are:
+
+``include-subprojects:true``
+  This doesn't make it clear from what
+  projects we are going to include subprojects from.
+  We could make it so it returns subprojects for all projects.
+  Users will probably use this with one project only.
+
+``subprojects:project/version`` (inclusive)
+  This allows to specify from what project exactly
+  we are going to return subprojects from,
+  and also include the version we are going to try to match.
+  This includes the parent project in the results.
+
+  As the ``project`` parameter, the version can be optional,
+  and defaults to the default version of the parent project.
+
+``subprojects:project/version`` (exclusive)
+  This is the same as the above,
+  but it doesn't include the parent project in the results.
+  If we want to include the results from the project, then
+  the query will be ``project:project/latest subprojects:project/latest``.
```

Comment:
I like this pattern. It's pretty explicit what the user wants:

* `project:docs/latest`: won't include subprojects at all
* `subprojects:docs`: will include _only_ results from all the subprojects from the `docs` project
* `project:docs/latest subprojects:docs`: will include results from `docs` _and_ all its subprojects

BTW, I would name it `subproject` (singular) to match the other key `project` and avoid confusions.